_source/_posts/2021-12-16-cloud-native-java-microservices-with-istio.md

```diff
@@ -0,0 +1,580 @@
+
+```bash
+$ cd istio-$ISTIO_VERSION
+$ kubectl apply -f samples/addons/grafana.yaml
+$ kubectl apply -f samples/addons/prometheus.yaml
+$ kubectl apply -f samples/addons/kiali.yaml
+$ kubectl apply -f samples/addons/extras/zipkin.yaml
+```
+
+{% img blog/cloud-native-java-microservices-with-istio/istio-pods.png alt:"GKE Cluster with Istio pods" width:"900" %}{: .center-image }
+
+If we look at the istio-system namespace, we can see all the Istio components along with Grafana, Prometheus, Kiali and Zipkin running. You can also see this by running `kubectl get pods -n istio-system`.
+
+## Deploy the microservice stack to GKE
+
+Our cluster is ready and we have Istio installed. Now, we can deploy our microservice stack to the cluster. But before that we need to create Kubernetes manifests for our deployments and services and configurations for Istio. And once again JHipster comes to the rescue. We can use the [JDL deployment](https://www.jhipster.tech/jdl/deployments) configurations to easily generate Kubernetes setup for our stack with one command.
+
+### Create Kubernetes manifests
+
+Create a new JDL file, say **deployment.jdl**, and add the following content to it.
+
+```kotlin
+// will be created under 'kubernetes' folder
+deployment {
+  deploymentType kubernetes
+  appsFolders [store, invoice, notification, product]
+  dockerRepositoryName "<your-docker-repository-name>"
+  serviceDiscoveryType no
+  istio true
+  kubernetesServiceType Ingress
+  kubernetesNamespace jhipster
+  ingressDomain "<istio-ingress-gateway-external-ip>.nip.io"
+  ingressType gke
+}
+```
+
+I hope its self explanatory. You can refer the JDL deployment documentation for all the available options. We have enabled Istio and set the ingress domain to the external IP of the Istio Ingress Gateway that we notes earlier. Make sure to use a docker repo where you have push rights.
+
+Now run the following command from the root folder where you ran the previous `jhipster jdl` command.
+
+```bash
+$ jhipster jdl deployment.jdl
+```
+
+This will create a new folder, **kubernetes**, with all the required Kubernetes manifests like deployments, services, Istio virtual services, gateways and so on for all the applications, databases, monitoring and so on.
+
+{% img blog/cloud-native-java-microservices-with-istio/k8s-deployment.png alt:"JHipster JDL deployment" width:"900" %}{: .center-image }
+
+As you can see, there is also a bunch of useful commands printed on the console that you can use to do the deployment.
+
+### Update OIDC configuration
+
+At this point we have everything ready, but since we didn't add the Okta OIDC client secret to the **application.yml** files, we need to add them as Kubernetes secrets. Under **kubernetes/store** folder add a new file names **okta-oidc-secrets.yml** with below content. You should ignore this file from GIT for security reasons.
```

Comment:
Change bold to backticks.

Reply:
GIT > git

Reply:
You can mention my post from last year that shows how to encrypt your secrets so you can check in these files.

https://developer.okta.com/blog/2021/06/01/kubernetes-spring-boot-jhipster#encrypt-your-kubernetes-secrets